packages/fxa-auth-server/scripts/check-users.js

```diff
@@ -0,0 +1,217 @@
+      const password = new Password(
+        credentials.authPW,
+        accountRecord.authSalt,
+        accountRecord.verifierVersion
+      );
+      const verifyHash = await password.verifyHash();
+      const passwordMatch = await this.db.checkPassword(
+        accountRecord.uid,
+        verifyHash
+      );
+
+      // Check to see if user has MFA enabled
+      let mfaEnabled = false;
+      try {
+        const totpToken = await this.db.totpToken(accountRecord.uid);
+        if (totpToken) {
+          mfaEnabled = true;
+        }
+      } catch (err) {}
+
+      const s = {
+        email: this.email,
+        exists: true,
+        passwordMatch,
+        mfaEnabled,
+        keysChangedAt: accountRecord.keysChangedAt,
+        profileChangedAt: accountRecord.profileChangedAt,
+        hasSecondaryEmails: accountRecord.emails.length > 1,
+        isPrimaryEmailVerified: accountRecord.primaryEmail.isVerified,
+      };
+      const stat = `${s.email},${s.exists},${sanitizeValue(
+        s.passwordMatch
+      )},${sanitizeValue(s.mfaEnabled)},${sanitizeValue(
+        s.keysChangedAt
+      )},${sanitizeValue(s.profileChangedAt)},${sanitizeValue(
+        s.hasSecondaryEmails
+      )},${sanitizeValue(s.isPrimaryEmailVerified)}`;
+
+      console.log(stat);
+
+      return s;
+    } catch (err) {
+      return {
+        exists: false,
+        email: this.email,
+      };
+    }
+  }
+}
+
+class CheckUsers {
+  constructor(filename) {
+    this.users = [];
+    this.db = undefined;
+    this.filename = filename;
+  }
+
+  getItems() {
+    try {
+      const input = fs
+        .readFileSync(path.resolve(this.filename))
+        .toString('utf8');
+
+      if (!input.length) {
+        return [];
+      }
+
+      // Parse the input file CSV style
+      return input.split(/\n/).map((s) => {
+        const delimiter = program.delimiter || ':';
+        const email = s.substring(0, s.indexOf(delimiter));
+        const password = s.substring(s.indexOf(delimiter) + 1, s.length);
+        return new User(email, password, this.db);
+      });
+    } catch (err) {
+      console.error('No such file or directory');
+      process.exit(1);
+    }
+  }
+
+  async load() {
+    this.db = await AuthDB.connect(config);
+    this.users = this.getItems();
+    console.info(
+      '%s accounts loaded from %s',
+      this.users.length,
+      this.filename
+    );
+  }
+
+  async close() {
+    await this.db.close();
+  }
+
+  async userStats() {
+    const stats = [];
+    for (const user of this.users) {
+      stats.push(await user.stats());
+    }
+    this.stats = stats;
+  }
+
+  saveStats() {
+    const stats = this.stats;
+    const output = [
+      'email,exists,passwordMatch,mfaEnabled,keysChangedAt,profileChangedAt,hasSecondaryEmails,isPrimaryEmailVerified',
+    ];
+
+    output.push(
+      ...stats.map((s) => {
+        return `${s.email},${s.exists},${sanitizeValue(
```

Comment:
Q: does `s.email` need to be escaped at all?
I'm not familiar with the live production data or how likely we are to have an `program.delimiter` (default: `:`) in the _local_ part of an email.
https://en.wikipedia.org/wiki/Email_address (which isn't a spec at all) seems to imply that:
> _"space and special characters `"(),:;<>@[\]` **are** allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in that quoted string, any backslash or double-quote must be preceded once by a backslash);"_

Or if we use some library if that handles escaping gotchas for us. I've used https://csv.js.org/ in the past, but admittedly only https://csv.js.org/parse/ 99% of the time and not too familiar w/ https://csv.js.org/stringify/.

Reply:
I thought about a dedicated library to do csv, but then was like meh to add another dependency. Production disallows the local part of an email to contain special characters. I think this is ok.

https://github.com/mozilla/fxa/blob/e06fcbeffab3af3e78be2fa6b8efa36bee7bb2f5/packages/fxa-auth-server/lib/routes/validators.js#L170